grpc/src/main/java/com/linecorp/armeria/server/grpc/UnframedGrpcService.java

```diff
@@ -273,31 +277,47 @@ private static void deframeAndRespond(
             }
         }
 
-        try (ArmeriaMessageDeframer deframer = new ArmeriaMessageDeframer(
-                new Listener() {
-                    @Override
-                    public void messageRead(DeframedMessage message) {
-                        // We know that we don't support compression, so this is always a ByteBuffer.
-                        final HttpData unframedContent = HttpData.wrap(message.buf()).withEndOfStream();
-                        unframedHeaders.setInt(HttpHeaderNames.CONTENT_LENGTH, unframedContent.length());
-                        res.complete(HttpResponse.of(unframedHeaders.build(), unframedContent));
-                    }
-
-                    @Override
-                    public void endOfStream() {
-                        if (!res.isDone()) {
-                            // If 'ResponseObserver.onCompleted()' is called without calling 'onNext()',
-                            // this callback would be invoked but 'messageRead' callback wouldn't.
-                            res.complete(HttpResponse.of(unframedHeaders.build()));
-                        }
-                    }
-                },
+        final ArmeriaMessageDeframerHandler handler = new ArmeriaMessageDeframerHandler(
                 // Max outbound message size is handled by the GrpcService, so we don't need to set it here.
-                Integer.MAX_VALUE,
-                ctx.alloc(), false)) {
-            deframer.request(1);
-            deframer.deframe(grpcResponse.content(), true);
-        }
+                Integer.MAX_VALUE);
+        final HttpDeframer<DeframedMessage> deframer = new HttpDeframer<>(handler, ctx.alloc());
+        StreamMessage.of(grpcResponse.content()).subscribe(deframer, ctx.eventLoop());
+        deframer.subscribe(singleSubscriber(unframedHeaders, res), ctx.eventLoop());
+    }
+
+    private static Subscriber<DeframedMessage> singleSubscriber(ResponseHeadersBuilder unframedHeaders,
+                                                                CompletableFuture<HttpResponse> res) {
+        return new Subscriber<DeframedMessage>() {
+
+            @Override
+            public void onSubscribe(Subscription subscription) {
+                subscription.request(1);
+            }
+
+            @Override
+            public void onNext(DeframedMessage message) {
+                // We know that we don't support compression, so this is always a ByteBuffer.
```

Comment:
nit: ByteBuf? ByteBuffer is the JDK library.

Reply:
It should be `ByteBuf`. 😅